sub/sd_sbr.c

```diff
@@ -0,0 +1,184 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <math.h>
+#include <limits.h>
+
+#include <subrandr/subrandr.h>
+
+#include "mpv_talloc.h"
+
+#include "options/m_config.h"
+#include "options/options.h"
+#include "common/common.h"
+#include "demux/packet_pool.h"
+#include "demux/stheader.h"
+#include "sub/osd.h"
+#include "video/mp_image.h"
+#include "sd.h"
+
+struct sd_sbr_priv {
+    struct sbr_library *sbr_library;
+    struct sbr_renderer *sbr_renderer;
+    struct sbr_subtitles *sbr_subtitles;
+    struct mp_osd_res osd;
+    struct sub_bitmaps *bitmaps;
+};
+
+static void enable_output(struct sd *sd, bool enable)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+    if (enable == !!ctx->sbr_renderer)
+        return;
+    if (ctx->sbr_renderer) {
+        sbr_renderer_destroy(ctx->sbr_renderer);
+        ctx->sbr_renderer = NULL;
+    } else {
+        ctx->sbr_renderer = sbr_renderer_create(ctx->sbr_library);
+    }
+}
+
+static int init(struct sd *sd)
+{
+    if (!sd->codec->codec)
+        return -1;
+    if (strcmp(sd->codec->codec, "textsub/srv3") && strcmp(sd->codec->codec, "textsub/vtt"))
+        return -1;
+
+    struct sd_sbr_priv *ctx = talloc_zero(sd, struct sd_sbr_priv);
+    sd->priv = ctx;
+
+    ctx->sbr_library = sbr_library_init();
+    ctx->bitmaps = talloc_zero(ctx, struct sub_bitmaps);
+    ctx->bitmaps->format = SUBBITMAP_BGRA;
+    ctx->bitmaps->num_parts = 1;
+    ctx->bitmaps->parts = talloc_zero(ctx->bitmaps, struct sub_bitmap);
+
+    enable_output(sd, true);
+
+    return 0;
+}
+
+static void decode(struct sd *sd, struct demux_packet *packet)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+
+    if (ctx->sbr_subtitles)
+        sbr_subtitles_destroy(ctx->sbr_subtitles);
+    if (ctx->sbr_renderer)
+        sbr_renderer_set_subtitles(ctx->sbr_renderer, NULL);
+
+    sbr_subtitle_format fmt = SBR_SUBTITLE_FORMAT_UNKNOWN;
+    if (!strcmp(sd->codec->codec, "textsub/srv3") )
+        fmt = SBR_SUBTITLE_FORMAT_SRV3;
+    else if (!strcmp(sd->codec->codec, "textsub/vtt"))
+        fmt = SBR_SUBTITLE_FORMAT_WEBVTT;
+
+    ctx->sbr_subtitles = sbr_load_text(
+        ctx->sbr_library,
+        packet->buffer,
+        packet->len,
+        fmt,
+        NULL
+    );
+    packet->sub_duration = packet->duration;
+}
+
+static struct sub_bitmaps *get_bitmaps(struct sd *sd, struct mp_osd_res dim,
+                                       int format, double pts)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+    struct mp_subtitle_opts *opts = sd->opts;
+
+    ctx->osd = dim;
+
+    if (pts == MP_NOPTS_VALUE || !ctx->sbr_renderer || !ctx->sbr_subtitles)
+        return NULL;
+
+    if (opts->sub_forced_events_only)
+        return NULL;
+
+    struct sbr_subtitle_context context = (sbr_subtitle_context) {
+        .dpi = 72,
```

Comment:
Ideally we would provide the real display DPI here instead of 72 to prevent rendering issues in the future if subrandr ever starts supporting absolute font sizes in subtitles (like WebVTT CSS). No idea how to get that here though.

Reply:
This likely should be passed the same way as `mp_osd_res` is passed as a argument thorough `render_object` and `sub_get_bitmaps`.

I would argue that you could extend `mp_osd_res` structure fro absolute scaling, if that something that is needed by this renderer.

Reply:
Ah that looks like a pretty invasive change, realistically I don't think anyone will actually even use absolute font sizes in WebVTT so it can be considered an edge case. I'm going to leave this for a follow up if it's ever deemed worthwhile.